src/libs/Authentication.js

```diff
@@ -0,0 +1,148 @@
+import requireParameters from './requireParameters';
+import * as Network from './Network';
+import * as NetworkStore from './Network/NetworkStore';
+import updateSessionAuthTokens from './actions/Session/updateSessionAuthTokens';
+import CONFIG from '../CONFIG';
+import redirectToSignIn from './actions/SignInRedirect';
+import CONST from '../CONST';
+import Log from './Log';
+
+/**
+ * @param {Object} parameters
+ * @param {Boolean} [parameters.useExpensifyLogin]
+ * @param {String} parameters.partnerName
+ * @param {String} parameters.partnerPassword
+ * @param {String} parameters.partnerUserID
+ * @param {String} parameters.partnerUserSecret
+ * @param {String} [parameters.twoFactorAuthCode]
+ * @param {String} [parameters.email]
+ * @param {String} [parameters.authToken]
+ * @returns {Promise}
+ */
+function Authenticate(parameters) {
+    const commandName = 'Authenticate';
+
+    requireParameters([
+        'partnerName',
+        'partnerPassword',
+        'partnerUserID',
+        'partnerUserSecret',
+    ], parameters, commandName);
+
+    return Network.post(commandName, {
+        // When authenticating for the first time, we pass useExpensifyLogin as true so we check
+        // for credentials for the expensify partnerID to let users Authenticate with their expensify user
+        // and password.
+        useExpensifyLogin: parameters.useExpensifyLogin,
+        partnerName: parameters.partnerName,
+        partnerPassword: parameters.partnerPassword,
+        partnerUserID: parameters.partnerUserID,
+        partnerUserSecret: parameters.partnerUserSecret,
+        twoFactorAuthCode: parameters.twoFactorAuthCode,
+        authToken: parameters.authToken,
+        shouldRetry: false,
+
+        // Force this request to be made because the network queue is paused when re-authentication is happening
+        forceNetworkRequest: true,
+
+        // Add email param so the first Authenticate request is logged on the server w/ this email
+        email: parameters.email,
+    })
+        .then((response) => {
+            // If we didn't get a 200 response from Authenticate we either failed to Authenticate with
+            // an expensify login or the login credentials we created after the initial authentication.
+            // In both cases, we need the user to sign in again with their expensify credentials
+            if (response.jsonCode !== 200) {
+                switch (response.jsonCode) {
+                    case 401:
+                        throw new Error('passwordForm.error.incorrectLoginOrPassword');
+                    case 402:
+                        // If too few characters are passed as the password, the WAF will pass it to the API as an empty
```

Comment:
NAB: This seems like something we should be preventing from the front-end. Do you know what the actual length of "too few characters" is?

Reply:
Would have to look up the exact number. I believe we prevent it on the front end as well (would have to check on that as well though).

Reply:
Looks like we are not https://github.com/Expensify/App/blob/84d058f5ade145a8bc89e00c36b340ba9078a8b0/src/pages/signin/PasswordForm.js#L84-L106 

Can create an issue for this.

Reply:
https://github.com/Expensify/App/issues/8794